### PR TITLE
feat: Add prompt injection protection mechanism

### PR DESCRIPTION
**Pull request by nextedoff**
## Purpose

This feature is adding a way to implement new protection methods for the query, focusing on prompt injection protection in this PR. For injection protection model, model named `protectai/deberta-v3-base-prompt-injection` was used.
New models and guard mechanisms can be added to `promptprotection.py`.

image
image

Currently, the injection protection can either be turned on via API and user interface as seen in the screenshot, or it can be set via environment variables:
`USE_INJECTION_PROTECTION="true"`

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

- [ ] Yes
- [x] No

## Does this require changes to learn.microsoft.com docs?

- [ ] Yes
- [x] No

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Code quality checklist

- [ ] The current tests all pass (`python -m pytest`).
- [ ] I added tests that prove my fix is effective or that my feature works
- [x] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

File: infra/main.parameters.json [hunk 1]
```diff
@@ -158,6 +158,9 @@
     "defaultModel": {
       "value": "${DEFAULT_MODEL}"
     },
+    "useInjectionProtection": {
+      "value": "${USE_INJECTION_PROTECTION}"
```
Comment (phoevos) [inline infra/main.parameters.json]: Let's also add the default here!

```suggestion
      "value": "${USE_INJECTION_PROTECTION=false}"
```

File: app/backend/api_wrappers/openai.py [hunk 2]
```diff
@@ -33,6 +33,9 @@ async def chat_completion(self, *args, **kwargs) -> ChatCompletion:
     async def create_embeddings(self, *args, **kwargs) -> CreateEmbeddingResponse:
         return await self.client.embeddings.create(*args, **kwargs)
 
+    async def text_classification(self, *args, **kwargs):
+        pass
```
Comment (phoevos) [inline app/backend/api_wrappers/openai.py]: Why don't we raise an error like we do in the HF wrapper?

```suggestion
        raise NotImplementedError
```

What will happen if `check_for_injection` runs with an OpenAI client?

File: app/backend/api_wrappers/hugging_face.py [hunk 3]
```diff
@@ -74,6 +75,11 @@ async def chat_completion(
             top_p=top_p,
         )
 
+    async def text_classification(
+        self, text: str, model: Optional[str] = None
```
Comment (phoevos) [inline app/backend/api_wrappers/hugging_face.py]: Just noticed that we do the same for the chat completion API, but why is the model optional? What happens if we don't provide it?

File: app/backend/app.py [hunk 4]
```diff
@@ -449,6 +452,8 @@ async def setup_clients():
     HUGGINGFACE_API_KEY = os.getenv("HUGGINGFACE_API_KEY")
     DEFAULT_MODEL = os.getenv("DEFAULT_MODEL")
 
+    USE_INJECTION_PROTECTION = os.getenv("USE_INJECTION_PROTECTION", "false").lower() == "true"
```
Comment (phoevos) [inline app/backend/app.py]: Isn't it this way?

```suggestion
    USE_INJECTION_PROTECTION = os.getenv("USE_INJECTION_PROTECTION", "").lower() == "true"
```

Not that it won't work, but everywhere else we do it the other way.

File: app/backend/app.py [hunk 5]
```diff
@@ -604,10 +609,15 @@ async def setup_clients():
         hf_client = HuggingFaceClient(token=HUGGINGFACE_API_KEY)
 
     llm_clients: dict[str, LLMClient] = {"openai": openai_client, "hf": hf_client}
+    protection_config = PromptProtectionConfig()
+
+    if USE_INJECTION_PROTECTION:
+        protection_config.injection_protection.use = True
```
Comment (phoevos) [inline app/backend/app.py]: This feels a bit wordy. Can't we do something like this?

```suggestion
    protection_config = PromptProtectionConfig(use=USE_INJECTION_PROTECTION)
```

Comment (phoevos) [reply app/backend/app.py]: Oh no, we can't due to the way these dataclasses are modelled. I'll have to think a bit about this but it does look weird exposing an API like this.

File: app/backend/core/promptprotection.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    def getModelName(self, protection_name: str) -> str:
+        return getattr(self, protection_name).model_name
+
+
+class PromptProtection:
+    def __init__(self, config: Optional[PromptProtectionConfig]):
+        if config:
+            self.config = config
+        else:
+            self.config = PromptProtectionConfig()
```
Comment (phoevos) [inline app/backend/core/promptprotection.py]:
```suggestion
        self.config = config if config else PromptProtectionConfig()
```

File: app/backend/error.py [hunk 7]
```diff
@@ -7,13 +8,22 @@
...
+    code: Optional[str] = None
+
+    def __init__(self, message: str, code: Optional[str]) -> None:
+        super().__init__(message)
+        self.code = code
 
 
 def error_dict(error: Exception) -> dict:
-    if isinstance(error, APIError) and error.code == "content_filter":
+    if (isinstance(error, APIError) or isinstance(error, PromptProtectionError)) and error.code == "content_filter":
```
Comment (phoevos) [inline app/backend/error.py]: Check out this cool syntactic sugar for what you're doing there (passing a tuple of types to check against)!
```suggestion
    if isinstance(error, (APIError, PromptProtectionError)) and error.code == "content_filter":
```

File: app/backend/error.py [hunk 8]
```diff
@@ -22,6 +32,6 @@ def error_dict(error: Exception) -> dict:
 
 def error_response(error: Exception, route: str, status_code: int = 500):
     logging.exception("Exception in %s: %s", route, error)
-    if isinstance(error, APIError) and error.code == "content_filter":
+    if (isinstance(error, APIError) or isinstance(error, PromptProtectionError)) and error.code == "content_filter":
```
Comment (phoevos) [inline app/backend/error.py]: Check out this cool syntactic sugar for what you're doing there (passing a tuple of types to check against)!
```suggestion
    if isinstance(error, (APIError, PromptProtectionError)) and error.code == "content_filter":
```

**Review by phoevos — CHANGES_REQUESTED** [review]
Thanks for opening this!

Left some comments with suggestions, but I'm not finished yet. I'll continue tomorrow to check the core of the implementation and the frontend changes (which I'm duly excited about)!

File: app/backend/api_wrappers/openai.py (hunk 2 above)
Comment (nextedoff) [reply app/backend/api_wrappers/openai.py]: Well, it was not expected that we would do something similar with the OpenAI client, and it's hardcoded to work with the Hugging Face Hub client.

File: app/backend/api_wrappers/hugging_face.py (hunk 3 above)
Comment (nextedoff) [reply app/backend/api_wrappers/hugging_face.py]: That's my mistake, I was also planning to make the PromptProtection instance to store its own hugging face client, which would allow to set the default model to use, then this would be optional.

File: app/backend/api_wrappers/openai.py (hunk 2 above)
Comment (phoevos) [reply app/backend/api_wrappers/openai.py]: Should this be clearer in the code? The OpenAI one supposedly has built-in protections for stuff like that, right?

Comment (nextedoff) [reply app/backend/api_wrappers/openai.py]: > Should this be clearer in the code? The OpenAI one supposedly has built-in protections for stuff like that, right?

Yeah, larger models (including gpt3.5-4) usually have guardrails for that.
I will add the notimplemented exception.

File: infra/main.parameters.json (hunk 1 above)
Comment (nextedoff) [reply infra/main.parameters.json]: Done

File: app/backend/api_wrappers/openai.py (hunk 2 above)
Comment (nextedoff) [reply app/backend/api_wrappers/openai.py]: Done

File: app/backend/api_wrappers/hugging_face.py (hunk 3 above)
Comment (nextedoff) [reply app/backend/api_wrappers/hugging_face.py]: Fixed

File: app/backend/app.py (hunk 4 above)
Comment (nextedoff) [reply app/backend/app.py]: fixed

File: app/backend/error.py (hunk 7 above)
Comment (nextedoff) [reply app/backend/error.py]: Done

File: app/backend/error.py (hunk 8 above)
Comment (nextedoff) [reply app/backend/error.py]: Done

File: app/backend/app.py [hunk 9]
```diff
@@ -605,9 +610,13 @@ async def setup_clients():
 
     llm_clients: dict[str, LLMClient] = {"openai": openai_client, "hf": hf_client}
 
+    prompt_protection = PromptProtection(prompt_protection_enabled=USE_INJECTION_PROTECTION)
+    print(prompt_protection)
```
Comment (phoevos) [inline app/backend/app.py]: Likely a debugging statement, let's drop it!
```suggestion
```

**Review by phoevos — APPROVED** [review]
I'll do some quick manual testing, but other than that LGTM!
